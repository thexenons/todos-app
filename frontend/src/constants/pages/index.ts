import { lazy } from "react";

import { PageKey, Pages } from "../../types";

export const PAGES: Pages = {
	[PageKey.home]: {
		path: "/",
		title: "Home",
		component: lazy(() => import("../../pages/Home")),
	},
	[PageKey.login]: {
		path: "/login",
		component: lazy(() => import("../../pages/Login")),
	},
	[PageKey.register]: {
		path: "/register",
		component: lazy(() => import("../../pages/Register")),
	},

	[PageKey.test]: {
		path: "/test",
		title: "Test",
		loader: async () => {
			return await (await import("../../pages/Test")).getInitialData();
		},
		component: lazy(() => import("../../pages/Test")),
		isProtected: true,
		children: {
			[PageKey.todosListSingle]: {
				path: "/:id",
				title: "Todos List Single",
				loader: async (args) => {
					return await (
						await import("../../pages/Test/TodosListSingle")
					).getInitialData(args);
				},
				component: lazy(() => import("../../pages/Test/TodosListSingle")),
			},
		},
	},
	[PageKey.components]: {
		path: "/components",
		title: "Components",
		component: lazy(() => import("../../pages/Components")),
		hideLayout: true,
		children: {
			[PageKey.componentsButton]: {
				path: "/button",
				title: "Button",
				component: lazy(() => import("../../pages/Components/Button")),
			},
			[PageKey.componentsAcceptButton]: {
				path: "/accept-button",
				title: "AcceptButton",
				component: lazy(() => import("../../pages/Components/AcceptButton")),
			},
			[PageKey.componentsUpdateButton]: {
				path: "/update-button",
				title: "UpdateButton",
				component: lazy(() => import("../../pages/Components/UpdateButton")),
			},
			[PageKey.componentsCreateButton]: {
				path: "/create-button",
				title: "CreateButton",
				component: lazy(() => import("../../pages/Components/CreateButton")),
			},
			[PageKey.componentsDeleteButton]: {
				path: "/delete-button",
				title: "DeleteButton",
				component: lazy(() => import("../../pages/Components/DeleteButton")),
			},
			[PageKey.componentsCancelButton]: {
				path: "/cancel-button",
				title: "CancelButton",
				component: lazy(() => import("../../pages/Components/CancelButton")),
			},
			[PageKey.componentsPanel]: {
				path: "/panel",
				title: "Panel",
				component: lazy(() => import("../../pages/Components/Panel")),
			},
			[PageKey.componentsModal]: {
				path: "/modal",
				title: "Modal",
				component: lazy(() => import("../../pages/Components/Modal")),
			},
			[PageKey.componentsFloatingPanel]: {
				path: "/floating-panel",
				title: "FloatingPanel",
				component: lazy(() => import("../../pages/Components/FloatingPanel")),
			},
			[PageKey.componentsInputBase]: {
				path: "/input-base",
				title: "InputBase",
				component: lazy(() => import("../../pages/Components/InputBase")),
			},
		},
	},
};
