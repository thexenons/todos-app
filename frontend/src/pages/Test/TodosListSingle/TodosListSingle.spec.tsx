import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useLoaderData } from "react-router-dom";
import { vi } from "vitest";

import BaseTestComponent from "../../../tests/BaseTestComponent";
import { TEST_TODOS_LIST } from "../../../tests/constants";
import TodosListSingle, { getInitialData } from ".";

vi.mock("react-router-dom");

describe("TodosListSingle", () => {
	afterEach(vi.restoreAllMocks);

	it("Render", () => {
		vi.mocked(useLoaderData).mockReturnValue({ todos_list: TEST_TODOS_LIST });

		const { container } = render(
			<BaseTestComponent>
				<TodosListSingle />
			</BaseTestComponent>
		);

		expect(container).toMatchSnapshot();
	});

	it("Render without todos_lists", () => {
		vi.mocked(useLoaderData).mockReturnValue({});

		const { container } = render(
			<BaseTestComponent>
				<TodosListSingle />
			</BaseTestComponent>
		);

		expect(container).toMatchSnapshot();
	});

	it("onBack", async () => {
		vi.mocked(useLoaderData).mockReturnValue({ todos_list: TEST_TODOS_LIST });
		render(
			<BaseTestComponent>
				<TodosListSingle />
			</BaseTestComponent>
		);

		const clickableBack = screen.getByRole("button");
		expect(clickableBack).toBeInTheDocument();

		await userEvent.click(clickableBack);
	});
});

describe("getInitialData", () => {
	afterEach(fetchMock.mockClear);

	it("with id", async () => {
		fetchMock.mockOnce(JSON.stringify(TEST_TODOS_LIST));

		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		// @ts-ignore
		const data = await getInitialData({ params: { id: "1" } });

		expect(data).toEqual({
			todos_list: TEST_TODOS_LIST,
		});
	});

	it("without id", async () => {
		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		// @ts-ignore
		const data = await getInitialData({ params: {} });

		expect(data).toEqual({});
	});
});
