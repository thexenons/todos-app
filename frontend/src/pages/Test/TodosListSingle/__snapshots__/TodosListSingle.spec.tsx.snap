// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodosListSingle > Render 1`] = `<div />`;

exports[`TodosListSingle > Render without todos_lists 1`] = `<div />`;
